File: Arrays/GroupAnagram.py
from collections import defaultdict
input_list = ["eat", "tea", "tan", "ate", "nat", "bat"]

anagram_ans = defaultdict(list)

for item in input_list:
    print(sorted(item))
    sorted_str = tuple(sorted(item))
    # Tuple cuz you cannot use mutable interable datastructures as keys
    anagram_ans[sorted_str].append(item)
    print(sorted_str)

# anagram_ans = {}

# for item in input_list:
#     print(sorted(item))
#     sorted_str = tuple(sorted(item))
#     anagram_ans[sorted_str] = []
#     anagram_ans[sorted_str].append(item)
#     print(sorted_str)


print(anagram_ans.values())

File: Arrays/Java/groupAnagram.java
package Arrays.Java;

import java.util.HashMap;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class groupAnagram {
    public static void main(String[] args) {
        String[] input_list = { "eat", "tea", "tan", "ate", "nat", "bat" };
        System.out.println(anagrams(input_list));
    }

    public static List<List<String>> anagrams(String[] input_list) {
        HashMap<String, List<String>> map = new HashMap<>();
        for (int i = 0; i < input_list.length; i++) {
            char[] chars = input_list[i].toCharArray();
            Arrays.sort(chars);

            String sortedKey = new String(chars);

            map.putIfAbsent(sortedKey, new ArrayList<>());
            map.get(sortedKey).add(input_list[i]);

        }
        return new ArrayList<>(map.values());
    }
}
